docs/index.rst

```diff
@@ -1,260 +1,427 @@
 
-To facilitate this, Confit lets you *assign* to view objects using
-ordinary Python assignment. Assignment will add an overlay source that
-precedes all other configuration sources in priority. Here's an example
-of programmatically setting a configuration value based on a ``DEBUG``
-constant::
+  as_date(formats)
 
-    if DEBUG:
-        config['verbosity'] = 100
-    ...
-    my_logger.setLevel(config['verbosity'].get(int))
+::
 
-This example allows the constant to override the default verbosity
-level, which would otherwise come from a configuration file.
+  as_time(formats)
 
-Assignment works be creating a new "source" for configuration data at
-the top of the stack. This new source takes priority over all other,
-previously-loaded sources. You can cause this explicitly by calling the
-``set()`` method on any view. A related method, ``add()``, works
-similarly but instead adds a new *lowest-priority* source to the bottom
-of the stack. This can be used to provide defaults for options that may
-be overridden by previously-loaded configuration files.
+::
+
+  as_datetime(formats)
+
+::
+
+  as_filename
+
+Return a filename, substituting tildes and absolute-ifying relative paths.
+The filename is relative to the source that provided it. That is, a relative
+path in a file (including a Python module) is relative to the directory
+containing the file. A relative path in a command-line argument or an
+environment variable is relative to the current working directory.
+
+::
+
+  as_choice(choices)
+
+Assert the value is among the *choices* (using the ``in`` operator) and, if
+*choices* is a ``dict``, return the associated value.
+
+
+Custom casts
+^^^^^^^^^^^^
+
+It is easy to define your own casts. A cast is just a function that takes a
+stack of variables, in precedence order starting with most preferred, and
+returns a value or raises an exception.
+
+.. note:: The stack is necessary to implement as_filename as a function
+  separate from the Configuration interface. To support user-defined casts,
+  they must be separate from the interface, and to prevent built-in casts
+  from enjoying "privileged" status such that user-defined casts cannot do
+  all the same things, they must be separate from the interface as well.
+  Without the stack, as_filename cannot read the file origin of a variable to
+  perform relative path resolution. However, demanding that casts accept a
+  stack means that built-in types, e.g. int, bool, float, cannot be used as
+  casts; users have to remember to use as_int, as_bool, as_float. It might be
+  an acceptable trade-off.
+
+
+Miscellaneous
+-------------
+
+::
+
+  type_check(config, onerror="raise")
+
+Check that the type of the most preferred value (final override) matches the
+type of the least preferred value (default) for each variable. If a mismatch
+is found, take the action described in *onerror*, chosen from among the
+following:
+
+"raise" | Raise a ``TypeError``.
+"warn"  | Call ``logging.warning``.
```

Comment:
Sorry; that wasn't really clear. I just meant that it might be useful to do:

    config['var'].get(type_check)

or something similar to ensure that the true value matches the type of the default for `var`.